tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,703 @@
+    ///    lock is held by the driver itself.
+    ///
+    /// It follows that cancellation of a timer requires acquiring the driver
+    /// lock if the state is Registered or FiringInProgress.
+    state: state_cell::StateCell,
 
-        self.waker.wake();
-    }
+    /// AtomicWaker for notifying the owning thread. May be updated without
+    /// holding locks.
+    ///
+    /// There is a risk of a race between the driver updating the timer state
+    /// and the waker being fired. To resolve this, we only guarantee that the
+    /// waker will be fired if the state can be observed to be Registered
+    /// _after__ setting the waker. In particular, if FiringInProgress is
+    /// observed, we do not guarantee that the waker will be invoked (but it may
+    /// be invoked).
+    waker: AtomicWaker,
 
-    pub(crate) fn error(&self, error: Error) {
-        // Record the precise nature of the error, if there isn't already an
-        // error present. If we don't actually transition to the error state
-        // below, that's fine, as the error details we set here will be ignored.
-        self.error.compare_and_swap(0, error.as_u8(), SeqCst);
+    /// Cache-padded contents (data mostly manipulated by the IO driver). This
+    /// is separated to avoid contention between threads polling the timer and
+    /// the driver thread itself.
+    padded: CachePadded<TimerSharedPadded>,
 
-        // Only transition to the error state if not currently elapsed
-        let mut curr = self.state.load(SeqCst);
+    _p: PhantomPinned,
+}
 
-        loop {
-            if is_elapsed(curr) {
-                return;
-            }
+impl TimerShared {
+    /// Gets the cached time-of-expiration value with relaxed memory order
+    pub(super) fn cached_when(&self) -> u64 {
+        self.padded.0.cached_when.load(Ordering::Relaxed)
+    }
 
-            let next = ERROR;
+    /// Gets the true time-of-expiration value, and copies it into the cached
+    /// time-of-expiration value.
+    ///
+    /// SAFETY: Must be called with the driver lock held, and when this entry is
+    /// not in any timer wheel lists.
+    pub(super) unsafe fn sync_when(&self) -> u64 {
+        let true_when = self.true_when();
 
-            let actual = self.state.compare_and_swap(curr, next, SeqCst);
+        self.padded
+            .0
+            .cached_when
+            .store(true_when, Ordering::Relaxed);
 
-            if curr == actual {
-                break;
-            }
+        true_when
+    }
+
+    /// Returns the true time-of-expiration value, with relaxed memory ordering.
+    pub(super) fn true_when(&self) -> u64 {
+        self.padded.0.true_when.load(Ordering::Relaxed)
+    }
+
+    /// Sets the true time-of-expiration value, with relaxed memory ordering.
+    /// Returns the old value.
+    pub(super) fn set_when(&self, t: u64) -> u64 {
+        self.padded.0.true_when.swap(t, Ordering::Relaxed)
```

Comment:
on x86 this introduces an unnecessary acq/rel fence as it uses the `xchgq` operation (or on x86-32 `cmpxchg8b`), this should be replaced with a simple load/store as there is only one writer